### PR TITLE
WIP: improve support for crosscompiling to musl32

### PR DESCRIPTION
**Pull request by wfranzini**
###### Motivation for this change

Crosscompiling packages to musl32 lead to compilation errors due to the missing `__stack_chk_fail_local` symbol.  Even building nix fails. This PR disable stackprotector for musl32.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -343,7 +343,9 @@ stdenv.mkDerivation {
       hardening_unsupported_flags+=" stackprotector pic"
     '' + optionalString (targetPlatform.libc == "newlib") ''
       hardening_unsupported_flags+=" stackprotector fortify pie pic"
-    '' + optionalString targetPlatform.isNetBSD ''
+      '' + optionalString (targetPlatform.libc == "musl" && targetPlatform.isi686) ''
```
Comment (flokli) [inline pkgs/build-support/cc-wrapper/default.nix]: is this only a problem for i686, or 32 bit in general?

Comment (wfranzini) [reply pkgs/build-support/cc-wrapper/default.nix]: musl32 is declared as i686-unknown-linux-musl

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix]: I got the same issue on i486 and i586 as well. This should be ```targetPlatform.isx86_32``` to cover those cases.

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix]: Make sure spacing is consistent here as well.

File: pkgs/development/libraries/libexecinfo/default.nix [hunk 2]
```diff
@@ -20,14 +20,11 @@ stdenv.mkDerivation rec {
       url = https://git.alpinelinux.org/cgit/aports/plain/main/libexecinfo/20-define-gnu-source.patch?id=730cdcef6901750f4029d4c3b8639ce02ee3ead1;
       sha256 = "1mp8mc639b0h2s69m5z6s2h3q3n1zl298j9j0plzj7f979j76302";
     })
-    (fetchpatch {
-      name = "30-linux-makefile.patch";
-      url = https://git.alpinelinux.org/cgit/aports/plain/main/libexecinfo/30-linux-makefile.patch?id=730cdcef6901750f4029d4c3b8639ce02ee3ead1;
-      sha256 = "1jwjz22z5cjy5h2bfghn62yl9ar8jiqhdvbwrcfavv17ihbhwcaf";
-    })
+    ./30-linux-makefile.patch
```
Comment (matthewbauer) [inline pkgs/development/libraries/libexecinfo/default.nix]: I wonder if alpine has a way to get ssp working?

Comment (wfranzini) [reply pkgs/development/libraries/libexecinfo/default.nix]: @matthewbauer Alpinelinux creates a small libssp_nonshared library in the musl-dev package.
https://git.alpinelinux.org/aports/tree/main/musl?h=master

I've recreated it in my local nixpkgs repo, however I'm not sure how to automatically link with libssp_nonshared.a.

Do you have a hint?

Comment (matthewbauer) [reply pkgs/development/libraries/libexecinfo/default.nix]: You probably need to put it in musl's $out/lib/libssp_nonshared.a. Unfortunately we also need to apply this patch to GCC:

https://git.alpinelinux.org/aports/tree/main/gcc/gcc-6.1-musl-libssp.patch

That would probably be okay to apply to GCC, but maybe only on i686-musl builds.

**Comment by 7c6f434c** [comment]
For context: is this problem irrelevant in 64-bit case?

**Comment by matthewbauer** [comment]
> For context: is this problem irrelevant in 64-bit case?

I have never seen it on x86_64. I think it could be possible to hit it with some messy linking, but I would prefer to limit the patching to just i*86 for now.

**Comment by symphorien** [comment]
I successfully used this PR to compile openssl.

**Comment by Ericson2314** [comment]
Can we submit some of these patches upstream and `fetchpatch` them?

**Comment by wfranzini** [comment]
> Can we submit some of these patches upstream and `fetchpatch` them?

@Ericson2314  I'll look into this the next week.  

I would like to have this PR merged somewhere in the future, but I need guidance about the correct branch to use since I fear it will cause a massive rebuild.

**Comment by flokli** [comment]
@wfranzini this should target the `staging` branch.

**Comment by symphorien** [comment]
I would really like to see this merged in one form or another.

File: pkgs/os-specific/linux/audit/audit-strndupa-rawmemchr-compat.patch [hunk 3]
```diff
@@ -0,0 +1,141 @@
+From d579a08bb1cde71f939c13ac6b2261052ae9f77e Mon Sep 17 00:00:00 2001
+From: Steve Grubb <sgrubb@redhat.com>
+Date: Tue, 26 Feb 2019 18:33:33 -0500
+Subject: [PATCH] Add substitue functions for strndupa & rawmemchr
```
Comment (7c6f434c) [inline pkgs/os-specific/linux/audit/audit-strndupa-rawmemchr-compat.patch:4]:
```suggestion
Subject: [PATCH] Add substitute functions for strndupa & rawmemchr
```

**Review by 7c6f434c — APPROVED** [review]

**Comment by wfranzini** [comment]
Later today, I'm going to open a new PR starting from staging.

**Comment by wfranzini** [comment]
I'll continue working on the cross compilation to musl32 in PR #89471

**Pull request closed without merging**